### DISCUSSION TITLE
Question about package dependencies vs. vcpkg_find_acquire_program

### DISCUSSION THREAD

**Discussion by bradlitterell** [microsoft/vcpkg#26279] [Q&A]
I notice a lot of packages use `vcpkg_find_acquire_program(NASM)`, but NASM does not show up as a top-level package in `vcpkg search`.  

I'm not sure what this means for me if I want to install and use NASM.  Is the best practice to install some other package I don't need in order to acquire the tool?

Coming from Linux and `apt-get`, my expectation was that package dependencies would simply get resolved so any package needing NASM would simply have a dependency on it, and it would be a top-level package that could be installed on it's own. 

I noticed the same thing regarding Strawberry-Perl.

Apologies if this is answered somewhere in the documentation already, I didn't find it in a few minutes of searching.

Thanks,
Brad

**Comment by ras0219-msft** [comment]
`vcpkg_find_acquire_program()`[1] is a helper function, not a port dependency. Unlike system-wide managers like apt, vcpkg ports are specialized for the problem of C++ library management and generally do not represent applications (with certain exceptions). Port dependencies are listed in the `"dependencies"`[2] field of a `vcpkg.json` file.

The `vcpkg_find_acquire_program(...)` helper instructs vcpkg to find or download one of a small select list of tools commonly required to build OSS packages, such as NASM. By making it a helper that runs as part of the build, instead of a port dependency, we can avoid downloading NASM entirely in the happy path of 100% hits on binary caching[3].

Today, vcpkg is not intended to be a method for you to acquire NASM for your own build -- when consuming from vcpkg, NASM is part of your "toolchain" (like the C++ compiler) that you should acquire using your existing methods (Visual Studio / winget / apt / chocolatey / etc). We internally acquire it in order to encapsulate the requirements of other libraries, so you only need to acquire the tools your own code needs and not all the tools needed for all the libraries you want to use.

However, we are looking at making vcpkg more capable in that area; please keep an eye on future developments of vcpkg artifacts[4]!

[1]: https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_find_acquire_program.cmake
[2]: https://github.com/microsoft/vcpkg/blob/581099d541ec960db5c2a8efe4639f1e2e25de60/ports/cpprestsdk/vcpkg.json#L11
[3]: https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md
[4]: https://devblogs.microsoft.com/cppblog/vcpkg-april-2022-release-is-now-available/#vcpkg-artifacts-merged-into-microsoft-vcpkg-tool-repo

**Reply by bradlitterell** [reply]
You described what acquire_program _does_.  I understood that from reading it.  

>I'm not sure what this means for me if I want to install and use NASM. Is the best practice to install some other package I don't need in >order to acquire the tool?

What I don't understand is this:

1. Does VCPkg resolve dependencies between packages?  If I need X that depends on Y, can I get X from VCPkg and have it automatically install Y?  [ I was assuming the answer to this is yes, but I'm no longer sure since tools (which are a dependency) get installed in a special way. ]

2. I'm creating a cross-platform OSS library.  I need a tool to build it.  On Linux I can get the tool via apt-get.
On Windows I get the tool how?? From a secure supply chain perspective I've been told the safe/approved way to consume OSS is to use VCPkg, but the only way to get the tools I need (NASM & Strawberry Perl) from VCPkg is to install _another_ package I don't need.

Do I understand that correctly?
And, if so, _should_ that be the primary answer for acquiring tools? Or should they be first-class ports within VCPkg instead?

**Reply by ras0219-msft** [reply]
1. Yes, vcpkg follows dependencies between packages.

2. From your perspective as a consumer, vcpkg manages OSS _library_ dependencies for your application. vcpkg is not (currently) a method to get a C++ compiler, a buildsystem, an editor, an ide, or other "independent tools". As mentioned above, we're investigating that space under the name "vcpkg artifacts". To get NASM or Perl on Windows, we recommend WinGet or another application management tool approved by your enterprise.
